library(BiocManager)
BiocManager::install(c("ATACseqQC", "ChIPpeakAnno", "MotifDb", "GenomicAlignments",
                       "BSgenome.Hsapiens.UCSC.hg19", "TxDb.Hsapiens.UCSC.hg19.knownGene",
                       "phastCons100way.UCSC.hg19", "ggplot2"))
library(ATACseqQC)
library(ggplot2)

source(system.file("extdata", "IGVSnapshot.R", package = "ATACseqQC"))

bamfile <- "/n/data1/dfci/pedonc/kadoch/ben/projects/kevin_pou2f3/atac_analysis/alignment_results/bams/20230824_KSA072_NCIH1048_FHD609_rep3_4.sorted.bam"
shifted_bamfile <- bamfile <- "/n/data1/dfci/pedonc/kadoch/ben/projects/kevin_pou2f3/atac_analysis/alignment_results/bams/20230824_KSA072_NCIH1048_FHD609_rep3_4.s.shifted.bam"
figure_out <- "/n/data1/dfci/pedonc/kadoch/ben/projects/kevin_pou2f3/atac_analysis/test_atacseqqc"


bamfile.labels <- gsub(".bam", "", basename(bamfile))

# === Basic Prelim ===
estimateLibComplexity(readsDupFreq(bamfile))

fragSizeDist(bamfile, bamfile.labels)

# === Shifting to account for TN5 Insert Bias ===

possibleTag <- list("integer"=c("AM", "AS", "CM", "CP", "FI", "H0", "H1", "H2", 
                                "HI", "IH", "MQ", "NH", "NM", "OP", "PQ", "SM",
                                "TC", "UQ"), 
                    "character"=c("BC", "BQ", "BZ", "CB", "CC", "CO", "CQ", "CR",
                                  "CS", "CT", "CY", "E2", "FS", "LB", "MC", "MD",
                                  "MI", "OA", "OC", "OQ", "OX", "PG", "PT", "PU",
                                  "Q2", "QT", "QX", "R2", "RG", "RX", "SA", "TS",
                                  "U2"))
library(Rsamtools)
bamTop100 <- scanBam(BamFile(bamfile, yieldSize = 100),
                     param = ScanBamParam(tag=unlist(possibleTag)))[[1]]$tag
tags <- names(bamTop100)[lengths(bamTop100)>0]

## files will be output into outPath
outPath <- "splited"
dir.create(outPath)
## shift the coordinates of 5'ends of alignments in the bam file
library(TxDb.Hsapiens.UCSC.hg19.knownGene)
## if you don't have an available TxDb, please refer
## GenomicFeatures::makeTxDbFromGFF to create one from gff3 or gtf file.
seqinformation <- seqinfo(TxDb.Hsapiens.UCSC.hg19.knownGene)
info <- as(seqinformation, "GRanges")
gal <- readBamFile(bamfile, tag=tags, which=info, asMates=TRUE, bigFile=TRUE)
gal1 <- shiftGAlignmentsList(gal, outbam=shifted_bamfile)



# === Promoter/Transcript body (PT) score ===

library(TxDb.Hsapiens.UCSC.hg19.knownGene)
txs <- transcripts(TxDb.Hsapiens.UCSC.hg19.knownGene)
pt <- PTscore(gal1, txs)
plot(pt$log2meanCoverage, pt$PT_score, 
     xlab="log2 mean coverage",
     ylab="Promoter vs Transcript")

# === Nucleosome Free Regions (NFR) score ===

nfr <- NFRscore(gal1, txs)
plot(nfr$log2meanCoverage, nfr$NFR_score, 
     xlab="log2 mean coverage",
     ylab="Nucleosome Free Regions score",
     main="NFRscore for 200bp flanking TSSs",
     xlim=c(-10, 0), ylim=c(-5, 5))

# === Transcription Start Site (TSS) Enrichment Score ===

tsse <- TSSEscore(gal1, txs)
tsse$TSSEscore

plot(100*(-9:10-.5), tsse$values, type="b", 
     xlab="distance to TSS",
     ylab="aggregate TSS score")


# === Split Reads ===

library(BSgenome.Hsapiens.UCSC.hg19)
library(phastCons100way.UCSC.hg19)
## run program for chromosome 1 only
txs <- txs[seqnames(txs) %in% "chr1"]
genome <- Hsapiens
## split the reads into NucleosomeFree, mononucleosome, 
## dinucleosome and trinucleosome.
## and save the binned alignments into bam files.
objs <- splitGAlignmentsByCut(gal1, txs=txs, genome=genome, outPath = outPath,
                              conservation=phastCons100way.UCSC.hg19)
## list the files generated by splitGAlignmentsByCut.
dir(outPath)

# objs <- splitBam(bamfile, tags=tags, outPath=outPath,
#                  txs=txs, genome=genome,
#                  conservation=phastCons100way.UCSC.hg19)

# objs <- splitGAlignmentsByCut(gal1, txs=txs, outPath = outPath)










